### PR TITLE
Memory-Based Latent Initialization ($I_{mem}$) via Continuous Hopfield Network

### PR DESCRIPTION
**Pull request by yohannes4321**
This pull request introduces a new Hopfield-style associative memory module for memory-based initialization in predictive coding layers and integrates it throughout the model architecture. It also adds new configuration options for memory usage and updates the forward pass logic to support memory retrieval and updates. Additionally, it makes minor changes to evaluation and text generation routines to support the new memory features.
Screenshot 2026-03-08 214354
energy_plot (17)
perplexity_plot (16)

[Uploading FASTER PREDICTIVE CODING NETWORKS.pdf…]()

**Major additions and integrations:**

* Added a new `HopfieldMemory` module implementing a continuous Hopfield-style associative memory for use in predictive coding layers, with support for retrieval, bootstrapping, and local updates.
* Updated the `GPTConfig` class and model configuration logic to include new memory-related parameters: `use_memory_init`, `memory_slots`, `memory_delta`, and `memory_update_qk`. These parameters are now propagated throughout the codebase and have sensible defaults. [[1]](diffhunk://#diff-b617d6fe0f64dca7a1bc97079b7089118d246d0df43244f8d0dc375b6b05c983R45-R48) [[2]](diffhunk://#diff-4804420639531581f1d7982ddea2d5059d9556ede903d6a484191eb2bede9b76L53-R57) [[3]](diffhunk://#diff-bad475685268be936cf1fd5ae9c9f16e58587ca75bc13726ec1f87bcb4228211L53-R57) [[4]](diffhunk://#diff-0d966d6c14a6b78bc80ed395e2fe33e014e91bcaafdde33608359324e66a914bL133-R137) [[5]](diffhunk://#diff-c0d91a554859c3665d3cca854edc7a49502bd07aebede412e88f3acc9b15886fL106-R110)

**Model architecture changes:**

* Modified the initialization of predictive coding layers in `attention.py`, `embedding.py`, `mlp.py`, and `output.py` to accept and use the new memory-related configuration parameters.
* Enhanced the `PCTransformer` model's forward pass to compute and pass an observation tensor for memory retrieval to all predictive coding layers, and to control memory update behavior via a new `requires_update` argument.

**Evaluation and generation updates:**

* Updated evaluation and text generation routines to use the new `requires_update` argument when calling the model, ensuring correct memory update semantics during inference. )

These changes collectively add associative memory capabilities to the predictive coding model, enabling richer initialization and update strategies for the hidden states.This PR introduces memory-based latent-state initialization ($I_{mem}$) for the predictive-coding transformer. By integrating a continuous Hopfield associative memory, this update replaces weak (zero/random) latent initialization with observation-conditioned retrieval. This enhancement significantly improves convergence behavior for low-step predictive coding inference , aligning the architecture more closely with generative PC workflows where standard forward initialization ($I_{fw}$) is unavailable.

 **Mathematical Foundation**

    1 The Hopfield memory module utilizes a continuous query-key-value retrieval mechanism to approximate the converged hidden state distribution conditioned on the observation
    
    .Observation Construction: Observations are formed from input and target embeddings: $o = [x, y]$.
    
    2 Memory Retrieval: The initial latent state for a given layer is retrieved via the attention mechanism:
    
    
    3 Local Parameter Update (Memory Loss): The memory parameters are updated manually (bypassing the standard autograd optimizer) to minimize the squared-error mismatch between the retrieved state and the converged latent state at the final inference step
Screenshot 2026-03-08 205244

 What Was Addedhopfield_memory.py: A new continuous Hopfield-style associative memory module.Implements the core retrieval API (Q, K, V, b, delta).Includes memory_loss(...) computation.Handles one-time pseudo-inverse bootstrapping for the value matrix $V$ ($V = A^+ H_{ifw}$) to warm-start the memory.Contains local_update(...) for manual, local parameter updates based on the converged state, entirely avoiding backprop overhead.

**1 What Was Changed1. PC Layer Integration (pc_layer.py)**

- Removed the reliance on random/zero x_init for non-embedding latent initialization.
- Added memory configuration arguments (use_memory_init, memory_slots, memory_delta, memory_update_qk) to the PCLayer constructor.
- Introduced an internal memory registry (self.memories, self._obs_cache, self._memory_loss) and a register_memory(...) method
- .Initialization Behavior (init_x): Now accepts an observation tensor. If memory is enabled, it registers the memory, bootstraps $V$ on the first pass (or uses a neutral baseline), and retrieves $h^{(0)}$ from memory. Otherwise, it defaults to zeros.
- Update Behavior (forward): At the final inference step ($t = T-1$), if updates are enabled, the Hopfield memory is locally updated using the converged latent state $h^{(T)}$, and the resulting $\mathcal{L}_H$ is accumulated.

**2. Model-Level Wiring & Observation Construction (pc_t_model.py)**
Modified the forward signature from forward(self, target_ids, input_ids, use_kv_cache=False) to include requires_update=True.

Added observation construction in embedding space prior to PC initialization:
obs_x = word_embeddings(input_ids)
obs_y = word_embeddings(target_ids)
observation = cat([obs_x, obs_y], dim=-1)

Passed the observation tensor and the requires_update flag down through all PC sublayer initialization and forward calls.

3. Sublayer Config Propagation (embedding.py, attention.py, mlp.py, output.py)
Updated constructors across all architecture modules. Each PCLayer instantiation now explicitly receives the memory-related configuration values.

4. Config Schema & Pipeline Extension
config.py: Added core schema fields (use_memory_init, memory_slots, memory_delta, memory_update_qk).

- Runtime Control: Enforced strict inference-only behavior where appropriate to prevent parameter mutation.
- training.py explicitly uses requires_update=True.
- eval.py and generate_text.py use requires_update=False.
- Updated benchmark scripts (benchmark_T.py, benchmark_blocks.py) to parse and propagate the new memory configurations

How It Works (End-to-End Flow)

**- Observation Construction:** For each batch, the model builds a detached observation tensor $o$ representing the input and target in embedding space.
**- Memory-Based Initialization:** Each predictive coding layer (excluding the embedding cache path) queries its respective Hopfield memory using the observation to retrieve the optimal initial latent state $h^{(0)}$
**- .Inference:** The standard predictive coding inference loop runs for $T$ steps, executing local layer updates and local error propagation.
**- Memory Learning (Convergence):** At the final inference step ($T-1$), if requires_update=True, the converged latent state $h^{(T)}$ triggers a local update, pulling the memory parameters ($Q, K, V, b, \delta$) toward patterns that minimize future retrieval mismatch.Mode Control: Train mode allows mutation of both model and memory parameters. Eval/Generation mode disables these updates, ensuring deterministic, non-mutating runs.

### CODE REVIEW

**Comment by Saad-Musema** [comment]
@yohannes4321 I am closing this PR as it is on a branch where another experimental work is done and has other follow up commits to make sure it works properly. So please, if you have talked to @Betty987 and she has approved of the task, send a PR to another branch.
